src/par/README.md

```diff
@@ -15,16 +17,500 @@ TritonPart is an open-source constraints-driven partitioner.  It can be used to
+$thr_{ilp}$, ILP-based partitioning is performed. If not, a single round of 
+constraints-driven coarsening is conducted to further reduce the size of 
+$H_{overlay}$ and generate a coarser hypergraph $H_{overlay}^{'}$. Finally,
+multilevel refinement is performed to further improve the partitioning
+solution at each level of the hierarchy and return the improved solution
+$S^{'}$. 
+
+5. V-Cycle Refinement
+
+Cut-overlay clustering and partitioning produces a high-quality partitioning
+solution $S^{'}$. To improve it, there are three phases similar to *hMETIS*,
+namely: multilevel coarsening, ILP-based partitioning, and refinement. 
+Firstly, in multilevel partitioning, $S^{'}$ is used as a community guidance
+for the constraints-driven coarsening. Only vertices within the same block
+are permitted to be merged, to ensure that the current solution $S^{'}$
+is preserved in the coarsest hypergraph $H_{c_\xi}$. In the ILP-based 
+partitioning phase, if the number of vertices in $H_{c_\xi}$ does not 
+exceed $thr_{ilp}$, run ILP-based partitioning to improve $S^{'}$. Else,
+continue with $S^{'}$ in successive iterations of these two steps 
+(default set to 2). Refinement phase is conducted as per step 3. 
+
+![](./doc/algo.webp)
+<center>TritonPart algorithm at a glance</center>
+
+## Timing Aware Algorithm
+
+`par` can also be used as a timing-aware partitioning framework. A slack
+propagation methodology is used that optimizes cuts for both timing-critical
+and timing-noncritical paths. 
+
+1. Extraction of Timing Paths and Slack Information
+
+First, the top $P$
+timing-critical paths and the slack information of each hyperedge using 
+the wireload model (WLM) is obtained from *OpenSTA*. The timing cost of 
+each path is then calculated:
+
+$$t_p = (1- \frac{slack_p - \Delta}{clock\_period})^\mu$$
+
+where a fixed extra delay $\Delta$ is introduced for timing guardband,
+and $\mu$ (default 2) is the exponent. 
+
+The snaking factor of a path $SF(p)$ is defined as the maximum number of 
+block reentries along the path $p$. The timing cost of a hyperedge is 
+computed using the timing weight corresponding to hyperedge slack $slack_e$,
+and the accumulated timing cost of all paths traversing the hyperedge.
+
+$$t_e = (1- \frac{slack_e -\Delta}{clock\_period})^\mu + \sum_{\{p|e\in p\}}t_p$$
+
+2. Timing-aware Coarsening
+
+Timing-aware feature is achieved by adding a timing cost of hyperedge $t_e$
+to the connectivity score earlier mentioned. If vertices $(u,v)$ are
+associated with multiple critical paths, then they are more likely to be
+merged. This is reflected in the update score function:
+
+$$r_t(u,v) = \hat{r}(u,v) + \sum_{e\in\{I(v) \cap I(u)\}} \frac{\beta t_e}{|e| - 1}$$
+
+3. Timing-aware Refinement
+
+Timing-aware refinement is based on the similar cost function as the main
+algorithm. Instead, an additional slack propagation step is performed at 
+the end of each PM/FM/HER pass. 
```

Comment:
```suggestion
where $\rho$ is a normalization factor set to the average distance between two
vertex embeddings. When vertices $v_1, ... , v_t$ are merged into a single
vertex $v_{coarse}$, the corresponding vertex embedding $X_{v_{coarse}}$
is defined as the *center of gravity* of $t$ vertices:

$$X_{v_{coarse}} = \sum_{j=1}^{t} \frac{||w_{v_j}||}{M} X_{v_j},\ where\ M= \sum_{j=1}^t ||w_{v_j}|| $$

- **Community Guidance**: Only vertices within the same community are
  considered for merging.
- **Tie-breaking mechanism**: If multiple neighbor pairs have the same rating
  score, combine the lexicographically first unmatched vertex to break ties.

2. Initial Partitioning

After completing the coarsening process, an initial partitioning solution for
the coarsest hypergraph $H_c$ is derived. Two sub-steps are involved in this:
the best partitioning solution from random and VILE partitioning is chosen
from $\eta = 50$ runs as a warm-start to the ILP-based partitioner. The
optimization is based on only the cut size rather than the cost function
$\phi$ to keep the runtime reasonable.

3. Refinement

After a feasible solution $H_{c_\xi}$ is obtained by initial partitioning,
uncoarsening and move-based refinement is performed to improve the
partitioning solution. Three refinement heuristics are applied in sequence:
- **$K$-way pairwise FM (PM)**: This addresses multi-way partitioning
  as concurrent bi-partitioning problems in a restricted version of K-way
  Fiduccia–Mattheyses (FM) algorithm. First, $\lfloor K/2\rfloor $ pairs of
  blocks are obtained, with refinement-specific vertex movements restricted
  to associated paired blocks. Next, two-way FM is concurrently performed on
  all the block pairs. finally, a new configuration of block pairs is computed
  at the end of the PM.
- **Direct $K$-way FM**: Using $K$ priority queues, for each block $V_i$,
  establish a priority queue that stores the vertices that can be potentially
  moved from the current block to block $V_i$. This queue is ordered according
  to the gain of the vertices. Gain is defined as the reduction in cost
  function from the movement of the vertex from the current block to $V_i$.
  Next, after a vertex move, each priority queue is updated independently, thus
  enabling parallel updates via multi-threading. Next, early-stop is implemented
  by limiting the maximum number of vertices moved to 100 per pass. Finally,
  the *corking effect* is mitigated by traversing the priority queue belonging
  to the vertex with the highest gain and identifying a feasible vertex move.
- **Greedy Hyperedge Refinement (HER)**: First, randomly visit all
  hyperedges. For each hyperedge $e$ that crosses the partition boundary,
  determine whether a subset of vertices in $e$ can be moved without violating
  the multi-dimensional balance constraints. The objective is to make $e$
  entirely constrained in a block.

4. Cut-Overlay Clustering and Partitioning (COCP)

Cut-overlay Clustering and Partitioning (COCP) is a mechanism to
combine multiple good-quality partitioning solutions to generate an
improved solution. To begin, the sets of hyperedges cut in the $\theta$
candidate solutions are denoted as $E_1,..., E_\theta \subset E$.  First,
$\cup_{i=1}^\theta E_i$ is removed from the hypergraph $H(V, E)$, resulting in
a number of connected components. Next, all vertices within each connected
component are merged to form a coarser hypergraph $H_{overlay}$. If the
number of vertices in $H_{overlay}$ is less than $thr_{ilp}$, ILP-based
partitioning is performed. If not, a single round of constraints-driven
coarsening is conducted to further reduce the size of $H_{overlay}$
and generate a coarser hypergraph $H_{overlay}^{'}$. Finally, multilevel
refinement is performed to further improve the partitioning solution at
each level of the hierarchy and return the improved solution $S^{'}$.

5. V-Cycle Refinement

Cut-overlay clustering and partitioning produces a high-quality partitioning
solution $S^{'}$. To improve it, there are three phases similar to *hMETIS*,
namely multilevel coarsening, ILP-based partitioning, and refinement.
Firstly, in multilevel partitioning, $S^{'}$ is used as a community guidance
for the constraints-driven coarsening. Only vertices within the same block
are permitted to be merged to ensure that the current solution $S^{'}$
is preserved in the coarsest hypergraph $H_{c_\xi}$. In the ILP-based
partitioning phase, if the number of vertices in $H_{c_\xi}$ does not exceed
$thr_{ilp}$, run ILP-based partitioning to improve $S^{'}$. Otherwise,
continue with $S^{'}$ in successive iterations of these two steps (default
set to 2). The refinement phase is conducted as per step 3.

![](./doc/algo.webp)<center>TritonPart algorithm at a glance</center>

## Timing Aware Algorithm

`par` can also be used as a timing-aware partitioning framework. A slack
propagation methodology is used that optimizes cuts for both timing-critical
and timing-noncritical paths.

1. Extraction of Timing Paths and Slack Information

First, the top $P$ timing-critical paths and the slack information of each
hyperedge using the wireload model (WLM) is obtained from *OpenSTA*. The
timing cost of each path is then calculated:

$$t_p = (1- \frac{slack_p - \Delta}{clock\_period})^\mu$$

where a fixed extra delay $\Delta$ is introduced for timing guardband,
and $\mu$ (default 2) is the exponent.

The snaking factor of a path $SF(p)$ is defined as the maximum number
of block reentries along the path $p$. The timing cost of a hyperedge is
computed using the timing weight corresponding to hyperedge slack $slack_e$
and the accumulated timing cost of all paths traversing the hyperedge.

$$t_e = (1- \frac{slack_e -\Delta}{clock\_period})^\mu + \sum_{\{p|e\in p\}}t_p$$

2. Timing-aware Coarsening

The timing-aware feature is achieved by adding a timing cost of hyperedge
$t_e$ to the connectivity score earlier mentioned. If vertices $(u,v)$
are associated with multiple critical paths, then they are more likely to
be merged. This is reflected in the update score function:

$$r_t(u,v) = \hat{r}(u,v) + \sum_{e\in\{I(v) \cap I(u)\}} \frac{\beta t_e}{|e| - 1}$$

3. Timing-aware Refinement

Timing-aware refinement is based on a similar cost function as the main
algorithm. Instead, an additional slack propagation step is performed at
the end of each PM/FM/HER pass.
```